src/coord/src/sink_connector.rs

```diff
@@ -107,62 +117,65 @@ fn get_latest_ts(
         )
     })?;
 
-    let mut latest_message = None;
-    while let Some(message) = get_next_message(consumer, timeout)? {
-        latest_message = Some(message);
+    let (lo, hi) = consumer
+        .fetch_watermarks(consistency_topic, 0, timeout)
+        .map_err(|e| {
+            anyhow!(
+                "Failed to fetch metadata while reading from consistency topic: {}",
+                e
+            )
+        })?;
+
+    // Empty topic
+    if hi == 0 {
+        return Ok(None);
     }
 
-    if latest_message.is_none() {
-        // fetch watermarks to distinguish between a timeout reading end-1 and an empty topic
-        match consumer.fetch_watermarks(consistency_topic, 0, timeout) {
-            Ok((lo, hi)) => {
-                if hi == 0 {
-                    return Ok(None);
-                } else {
-                    bail!(
-                        "uninitialized consistency topic {}:{}, lo/hi: {}/{}",
-                        consistency_topic,
-                        partition,
-                        lo,
-                        hi
-                    );
-                }
-            }
-            Err(e) => {
-                bail!(
-                    "Failed to fetch metadata while reading from consistency topic: {}",
-                    e
-                );
+    let mut latest_ts = None;
+    let mut latest_offset = None;
+    while let Some((message, offset)) = get_next_message(&mut consumer, timeout)? {
+        debug_assert!(offset >= latest_offset.unwrap_or(0));
+        latest_offset = Some(offset);
+
+        if let Some(ts) = maybe_decode_consistency_end_record(&message, consistency_topic)? {
+            if ts >= latest_ts.unwrap_or(0) {
+                latest_ts = Some(ts);
             }
         }
     }
 
-    let latest_message = latest_message.expect("known to exist");
-
-    // the latest valid message should be an END message. If not, things have
-    // gone wrong!
-    let timestamp = decode_consistency_end_record(&latest_message, consistency_topic)?;
-
-    Ok(Some(timestamp))
+    // Topic not empty but we couldn't read any messages
+    if latest_offset.is_none() {
```

Comment:
I can't tell if this is _always_ an error. When we talked yesterday you said that you always read up to `high_watermark - 2` so, what if the high watermark was 2 or 1 -- is that a case where not reading any messages would be legit? It seems like maybe we could get rid of this error path unless I'm missing something

Reply:
Summary of chat in slack: we don't expect this to happen but we don't want to codify the expectation that kafka doesn't insert any internal messages (like it does after all of our END messages) at the beginning so we'll just log when this happen